File: july/baekjoon_2252.java
package day5;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class baekjoon_2252 {
	static int N,M;
	static int A,B;

	static int[] indegree;
	static ArrayList[] edge;
	static Queue<Integer> queue;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());

		// ArrayList 초기화
		edge = new ArrayList[N+1];
		for(int i=1; i<=N ; i++) {
			edge[i] = new ArrayList<Integer>();
		}

		// 초기화
		indegree = new int[N+1];

		M = Integer.parseInt(st.nextToken());
		for(int m=0; m<M; m++) {
			st = new StringTokenizer(br.readLine());

			A = Integer.parseInt(st.nextToken());
			B = Integer.parseInt(st.nextToken());

			// a < b
			edge[A].add(B);
			indegree[B] += 1;
		}

		// 진입차수 0인 얘들 Queue에 넣기
		queue = new LinkedList<Integer>();
		for(int i=1; i<=N; i++) {
			if(indegree[i] == 0)
				queue.add(i);
		}

		// Queue에서 1개씩 뽑으면서 연결된 간선 끊기
		StringBuilder sb = new StringBuilder();
		while(!queue.isEmpty()) {
			int now = queue.poll();

			sb.append(now + " ");

			int size = edge[now].size();
			for(int i=0; i<size; i++) {
				// 연결된 간선 찾기
				int target = (int)edge[now].get(i);

				// 해당 간선 끊기
				indegree[target]-=1;
				if(indegree[target]==0)
					queue.add(target);
			}
		}

		System.out.println(sb.toString());
	}

}

File: August/baekjoon_2252.py
# 백준 2252번 줄 세우기
# 위상 정렬
from collections import deque

N, M = map(int, input().split())

# 그래프 초기화
edge = [[] for _ in range(N+1)]
indegree = [0 for _ in range(N+1)]

for i in range(M):
    a, b = map(int, input().split())

    # a < b
    edge[a].append(b)
    indegree[b] += 1

# 진입 차수 0인 얘들 큐에 넣기
queue = deque()
for i in range(1, N+1):
    if indegree[i] == 0:
        queue.append(i)

# 큐에서 1개씩 뽑으면서 연결된 간선 끊기
while len(queue)>0:
    now = queue.popleft()
    print(now, end=' ')

    # 연결된 간선 끊기
    for t in edge[now]:
        indegree[t] -= 1
        if indegree[t] == 0:
            queue.append(t)
